src/main/scala/com/snowplowanalytics/snowplow/postgres/loader/streaming/sink.scala

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright (c) 2020 Snowplow Analytics Ltd. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package com.snowplowanalytics.snowplow.postgres.loader.streaming
+
+import cats.Monad
+import cats.data.EitherT
+import cats.implicits._
+
+import cats.effect.{Sync, Clock, Concurrent}
+import cats.effect.concurrent.Ref
+
+import fs2.Pipe
+import fs2.concurrent.Queue
+
+import doobie._
+import doobie.implicits._
+import doobie.util.transactor.Transactor
+
+import io.circe.Json
+
+import com.snowplowanalytics.iglu.core.circe.implicits._
+
+import com.snowplowanalytics.iglu.client.{Resolver, Client}
+
+import com.snowplowanalytics.snowplow.badrows.{BadRow, Failure, Payload}
+import com.snowplowanalytics.snowplow.postgres.loader.config.Cli.processor
+import com.snowplowanalytics.snowplow.postgres.loader.storage.{PgState, ddl}
+import com.snowplowanalytics.snowplow.postgres.loader.shredding.{Entity, transform}
+import com.snowplowanalytics.snowplow.postgres.loader.streaming.source.{Data, BadData}
+
+object sink {
+
+  type Insert = ConnectionIO[Unit]
+
+  def insertStatement(logger: LogHandler, schema: String, row: Entity): Insert = {
+    val length = row.columns.length
+
+    val columns = Fragment.const0(row.columns.map(c => s"""\"${c.name}\"""").mkString(","))
+
+    val table = Fragment.const0(s"$schema.${row.tableName}")
+    val values = row.columns.zipWithIndex.foldLeft(fr0"") {
+      case (acc, (cur, i)) if i < length - 1 => acc ++ cur.value.fragment ++ fr0","
+      case (acc, (cur, _)) => acc ++ cur.value.fragment
+    }
+
+    fr"""INSERT INTO $table ($columns) VALUES ($values)""".update(logger).run.void
+  }
+
+  /** Sink bad data coming directly into the `Pipe` and data coming from `badQueue` */
+  def badSink[F[_]: Concurrent](badQueue: Queue[F, BadData]): Pipe[F, BadData, Unit] =
+    _.merge(badQueue.dequeue).evalMap {
+      case BadData.BadEnriched(row) => Sync[F].delay(println(row.compact))
+      case BadData.BadJson(payload, error) => Sync[F].delay(println(s"Cannot parse $payload. $error"))
+    }
```

Comment:
If something goes wrong in the processing of one event (e.g. can't parse it, can't insert it), don't we want to send a `BadRow` to PubSub or Kinesis, depending on the platform?

Reply:
That's a good question. Generally - yes, but given its non-production status stdout will be just enough to figure out what's going on.